core/arch/arm/kernel/generic_entry_a32.S

```diff
@@ -156,26 +149,19 @@ UNWIND(	.fnstart)
 UNWIND(	.cantunwind)
 	mov	r4, r0		/* Save pageable part address */
 	mov	r5, lr		/* Save ns-entry address */
-	mov	r6, r2		/* Save DT address */
-
-#ifdef CFG_TEE_GDB_BOOT
-	/*
-	 * boot preloaded linux: save linux boot args from GDB
-	 * all following routines, until reset_primary is called,
-	 * preserve r6/r7/r8
-	 */
-	mov	r6, r0
-	mov	r7, r1
-	mov	r8, r2
+	mov	r6, r2		/* Save DT address (std bootarg#2) */
 
-#if !defined(CFG_BUILT_IN_ARGS) && (defined(CFG_WITH_PAGER) || defined (CFG_DT))
-	/*
-	 * r0/r2 are used for different purposes by GDB boot and pager/DT
-	 * so these configurations are incompatible. however, if built-in
-	 * argumements are used, r0/r2 are free for GDB boot use.
-	 */
-#error CFG_TEE_GDB_BOOT and (CFG_WITH_PAGER || CFG_DT) are incompatible
+#if defined(CFG_NS_ENTRY_ADDR)
+	ldr	r5, =CFG_NS_ENTRY_ADDR
+#endif
+#if defined(CFG_PAGEABLE_ADDR)
+	ldr	r4, =CFG_PAGEABLE_ADDR
 #endif
+#if defined(CFG_DT_ADDR)
+	ldr	r6, =CFG_DT_ADDR
+#endif
+#if defined(CFG_RELAY_STD_BOOTARGS)
+	mov	sp, r1		/* Save Boot argument (std bootarg #1) */
```

Comment:
OK, I'd rather not scratch `r7` in `copy_init` then. The performance decrease is probably not noticeable. And if it's truly a problem we could use `sp` as a scratch registers in `copy_init` instead.